### PR TITLE
Readonly members improved diagnostics

### PR DESCRIPTION
**Pull request by RikkiGibson**
- Commit 1 adds the following error codes:
```cs
        ERR_StaticMemberCantBeReadOnly = 8656,
        ERR_AutoPropertyCantBeReadOnly = 8657,
        ERR_InvalidPropertyReadOnlyMods = 8658,
        ERR_DuplicatePropertyReadOnlyMods = 8659,
        ERR_FieldLikeEventCantBeReadOnly = 8660
```
- Commit 2 updates the symbol classes to use these error codes
- Commit 3 improves errors for writable usage of readonly instance variables

cc @jcouv @agocke @333fred who have had questions or suggestions about what the language should do in some of these cases.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 1]
```diff
@@ -5427,6 +5427,36 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
   <data name="WRN_ImplicitCopyInReadOnlyMember_Title" xml:space="preserve">
     <value>Call to non-readonly member from a 'readonly' member results in an implicit copy.</value>
   </data>
+  <data name="ERR_StaticMemberCantBeReadOnly" xml:space="preserve">
+    <value>Static member '{0}' cannot be 'readonly'.</value>
+  </data>
+  <data name="ERR_StaticMemberCantBeReadOnly_Title" xml:space="preserve">
```
Comment (cston) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: >ERR_StaticMemberCantBeReadOnly_Title [](start = 14, length = 36)

Are the `ERR_..._Title` members used?

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/CSharpResources.resx]: I'd like to know that myself. I was copying the convention I saw for the error message resources in this file--maybe when you're filtering the error list in the IDE the title is used to display what each error is generally about?

Comment (cston) [reply src/Compilers/CSharp/Portable/CSharpResources.resx]: I believe `_Title` members are only needed for warnings, not errors.

---
In reply to: [267959720](https://github.com/dotnet/roslyn/pull/34338#discussion_r267959720) [](ancestors = 267959720)

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/CSharpResources.resx]: Removed.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs [hunk 2]
```diff
@@ -152,8 +152,8 @@ internal override bool IsExpressionBodied
             _isExpressionBodied = true;
 
             // The modifiers for the accessor are the same as the modifiers for the property,
-            // minus the indexer bit
-            var declarationModifiers = propertyModifiers & ~DeclarationModifiers.Indexer;
+            // minus the indexer and readonly bit
+            var declarationModifiers = propertyModifiers & ~(DeclarationModifiers.Indexer | DeclarationModifiers.ReadOnly);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs]: >propertyModifiers & ~(DeclarationModifiers.Indexer | DeclarationModifiers.ReadOnly) [](start = 39, length = 83)

Consider extracting a `static` helper method for use here and in constructor below.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs [hunk 3]
```diff
@@ -210,21 +210,15 @@ internal override bool IsExpressionBodied
...
                 // Private accessors cannot be virtual.
                 propertyModifiers &= ~DeclarationModifiers.Virtual;
             }
             declarationModifiers |= propertyModifiers & ~DeclarationModifiers.Indexer;
 
-            // auto-implemented struct getters are implicitly 'readonly'
-            if (containingType.IsStructType() && !property.IsStatic && isAutoPropertyAccessor && methodKind == MethodKind.PropertyGet)
-            {
-                declarationModifiers |= DeclarationModifiers.ReadOnly;
-            }
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs]: Why remove this and override `IsDeclaredReadOnly`? If we revert this, could we marked `SourceMemberMethodSymbol.IsDeclaredReadOnly` as `sealed`?

File: src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs [hunk 4]
```diff
@@ -462,10 +468,23 @@ private void CheckModifiers(Location location, bool hasBody, bool isAutoProperty
...
+            else if (LocalDeclaredReadOnly)
             {
-                // The modifier '{0}' is not valid for this item
-                diagnostics.Add(ErrorCode.ERR_BadMemberFlag, location, SyntaxFacts.GetText(SyntaxKind.ReadOnlyKeyword));
+                if (_property.HasReadOnlyModifier)
+                {
+                    // Cannot specify 'readonly' modifiers on both property or indexer '{0}' and its accessors.
+                    diagnostics.Add(ErrorCode.ERR_InvalidPropertyReadOnlyMods, location, _property);
+                }
+                if (IsStatic)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs]: >if [](start = 16, length = 2)

Is the `if` rather than `else if` intentional? Are we testing: a `static readonly` properties with `readonly` accessor; and a `readonly` auto-property with `readonly` accessor?

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs]: Oops, this should be `if`/`else if`/`else if`. Maybe this is an argument for flattening it out and doing repeated checks of `LocalDeclaredReadOnly`?

---
In reply to: [267975807](https://github.com/dotnet/roslyn/pull/34338#discussion_r267975807) [](ancestors = 267975807)

File: src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs (hunk 2 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs]: You mean a method like `GetAccessorModifiers(propertyModifiers)`? Then do the special private virtual tweak in the constructor taking AccessorDeclarationSyntax?

---
In reply to: [267972053](https://github.com/dotnet/roslyn/pull/34338#discussion_r267972053) [](ancestors = 267972053)

File: src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs [hunk 5]
```diff
@@ -1189,12 +1188,43 @@ public struct S
     public int P1 { readonly get; }
     public readonly int P2 { get; }
     public int P3 { readonly get; set; }
-    public int P4 { readonly get; readonly set; } // PROTOTYPE: readonly set on an auto-property should give an error
-    public readonly int P5 { get; set; } // PROTOTYPE: readonly set on an auto-property should give an error
+    public int P4 { readonly get; readonly set; }
+    public readonly int P5 { get; set; }
+    public readonly int P6 { readonly get; }
 }
 ";
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs]: For completeness, perhaps:
```C#
public int P7 { get; readonly set; }
public readonly int P8 { get; readonly set; }
```

File: src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs (hunk 2 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs]: Yes, use `GetAccessorModifiers(propertyModifiers)` in both, and keep the additional tweak in the second constructor.

---
In reply to: [267976876](https://github.com/dotnet/roslyn/pull/34338#discussion_r267976876) [](ancestors = 267976876,267972053)

File: src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs [hunk 6]
```diff
@@ -462,10 +468,23 @@ private void CheckModifiers(Location location, bool hasBody, bool isAutoProperty
...
             }
-            else if (IsStatic && IsDeclaredReadOnly && !_property.HasReadOnlyModifier)
+            else if (LocalDeclaredReadOnly)
             {
-                // The modifier '{0}' is not valid for this item
-                diagnostics.Add(ErrorCode.ERR_BadMemberFlag, location, SyntaxFacts.GetText(SyntaxKind.ReadOnlyKeyword));
+                if (_property.HasReadOnlyModifier)
+                {
+                    // Cannot specify 'readonly' modifiers on both property or indexer '{0}' and its accessors.
+                    diagnostics.Add(ErrorCode.ERR_InvalidPropertyReadOnlyMods, location, _property);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs]: >ErrorCode.ERR_InvalidPropertyReadOnlyMods [](start = 36, length = 41)

Are we testing `ERR_InvalidPropertyReadOnlyMods` or `ERR_StaticMemberCantBeReadOnly` below with indexers?

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs]: Looks like no. Will add the coverage.

---
In reply to: [267981220](https://github.com/dotnet/roslyn/pull/34338#discussion_r267981220) [](ancestors = 267981220)

File: src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs (hunk 5 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs]: Sounds good.

---
In reply to: [267979372](https://github.com/dotnet/roslyn/pull/34338#discussion_r267979372) [](ancestors = 267979372)

File: src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs [hunk 7]
```diff
@@ -1087,9 +1086,9 @@ public struct S
...
             var comp = CreateCompilation(csharp);
             comp.VerifyDiagnostics(
-                // (5,32): error CS0106: The modifier 'readonly' is not valid for this item
+                // (5,32): error CS8657: Static member 'S.M()' cannot be 'readonly'.
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs]: > error CS8657: Static member 'S.M()' cannot be 'readonly'. [](start = 26, length = 58)

Could you also test the symbol API (unless that's covered elsewhere), to illustrate that static members always answer `false` when asked if they are `readonly`?

Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs]: I think this file is a good place to add that.

---
In reply to: [267987488](https://github.com/dotnet/roslyn/pull/34338#discussion_r267987488) [](ancestors = 267987488)

File: src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs (hunk 6 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs]: Note that `StaticMemberCantBeReadOnly` can't be produced for indexers because indexers aren't allowed to be `static`.

File: src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs [hunk 8]
```diff
@@ -460,7 +460,7 @@ internal bool CheckValueKind(SyntaxNode node, BoundExpression expr, BindValueKin
                     if (!isValueType || (RequiresAssignableVariable(valueKind) && (this.ContainingMemberOrLambda as MethodSymbol)?.IsEffectivelyReadOnly == true))
                     {
                         // CONSIDER: the Dev10 name has angle brackets (i.e. "<this>")
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs]: While you're here, can you remove this? Clearly we aren't going to do this.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceEventSymbol.cs [hunk 9]
```diff
@@ -478,10 +478,15 @@ protected void CheckModifiersAndType(DiagnosticBag diagnostics)
                 // A static member '{0}' cannot be marked as override, virtual, or abstract
                 diagnostics.Add(ErrorCode.ERR_StaticNotVirtual, location, this);
             }
-            else if (IsReadOnly && (IsStatic || HasAssociatedField))
+            else if (IsReadOnly && IsStatic)
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceEventSymbol.cs]: I know you didn't choose how this chain of else-ifs was structured, but while you're here consider refactoring this so it's sane: ie, not checking `IsAbstract` 5 times, etc.

File: src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs [hunk 10]
```diff
@@ -75,9 +75,9 @@ void AssignField()
                 // (16,9): error CS1604: Cannot assign to 'this' because it is read-only
                 //         this = default;
                 Diagnostic(ErrorCode.ERR_AssgReadonlyLocal, "this").WithArguments("this").WithLocation(16, 9),
-                // (18,9): error CS1604: Cannot assign to 'this' because it is read-only
+                // (18,9): error CS1604: Cannot assign to 'this.x' because it is read-only
                 //         this.x = 1;
-                Diagnostic(ErrorCode.ERR_AssgReadonlyLocal, "this.x").WithArguments("this").WithLocation(18, 9));
+                Diagnostic(ErrorCode.ERR_AssgReadonlyLocal, "this.x").WithArguments("this.x").WithLocation(18, 9));
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs]: Beautiful

File: src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs [hunk 11]
```diff
@@ -1087,9 +1086,13 @@ public struct S
 ";
             var comp = CreateCompilation(csharp);
             comp.VerifyDiagnostics(
-                // (5,32): error CS0106: The modifier 'readonly' is not valid for this item
+                // (5,32): error CS8657: Static member 'S.M()' cannot be 'readonly'.
                 //     public static readonly int M()
-                Diagnostic(ErrorCode.ERR_BadMemberFlag, "M").WithArguments("readonly").WithLocation(5, 32));
+                Diagnostic(ErrorCode.ERR_StaticMemberCantBeReadOnly, "M").WithArguments("S.M()").WithLocation(5, 32));
+
+            var method = comp.GetMember<NamedTypeSymbol>("S").GetMethod("M");
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs]: Please also test the public API (unless you haven't hooked up the public version yet)

File: src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs [hunk 12]
```diff
@@ -1203,14 +1245,14 @@ public void ReadOnlyProperty_RedundantReadOnlyAccessor()
             var csharp = @"
 public struct S
 {
-    public readonly int P { readonly get; }
+    public readonly int P { readonly get => 42; }
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs]: > => 42 [](start = 40, length = 6)

Why did you need to add this?

File: src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs [hunk 13]
```diff
@@ -1617,45 +1659,56 @@ public void ReadOnlyIndexer()
...
+public struct S5
+{
+    // error
+    public static readonly int this[int i] => i;
+}
 ";
             var comp = CreateCompilation(csharp);
-            comp.VerifyDiagnostics();
+            comp.VerifyDiagnostics(
+                // (21,16): error CS8659: Cannot specify 'readonly' modifiers on both accessors of property or indexer 'S3.this[int]'.
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs]: Can we log a bug to add a quickfix to detect this and move them to the property/indexer outer part? Otherwise this just feels like the compiler is being obtuse.

**Comment by 333fred** [comment]
Done review pass (commit 5). Just some cleanup comments and test questions.

File: src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs (hunk 12 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs]: The prototype used to allow putting readonly on an auto property but the spec changed to disallow it.

File: src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs (hunk 13 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs]: Seems reasonable. Whatever is currently done for accessibility modifiers is probably a good baseline since that’s what this behavior is based on.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbol.cs [hunk 14]
```diff
@@ -177,11 +177,18 @@ internal sealed class SourcePropertySymbol : PropertySymbol, IAttributeTargetSym
...
 
-                if (_isAutoProperty && !isReadOnly && !IsStatic && ContainingType.IsReadOnly)
+                if (_isAutoProperty && !IsStatic)
                 {
-                    diagnostics.Add(ErrorCode.ERR_AutoPropsInRoStruct, location);
+                    if (!isGetterOnly && ContainingType.IsReadOnly)
+                    {
+                        diagnostics.Add(ErrorCode.ERR_AutoPropsInRoStruct, location);
+                    }
+                    else if (HasReadOnlyModifier)
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbol.cs]: Shouldn't this be conditioned on `!isGetterOnly`. A `readonly` modifier should be allowed on a `get` only property. It's just an explicit default.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs [hunk 15]
```diff
@@ -409,6 +405,18 @@ internal Accessibility LocalAccessibility
             get { return ModifierUtils.EffectiveAccessibility(this.DeclarationModifiers); }
         }
 
+        /// <summary>
+        /// Indicates whether this accessor itself has a 'readonly' modifier.
+        /// </summary>
+        internal bool LocalDeclaredReadOnly => (DeclarationModifiers & DeclarationModifiers.ReadOnly) != 0;
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs]: Prefix `bool` based properties with Is, Has, etc ...

Comment (jaredpar) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs]: Rather than local consider using "directly". It would be easy for local to get confused with locals. I think "direct" is harder to mis-interpret here. 

---
In reply to: [268237180](https://github.com/dotnet/roslyn/pull/34338#discussion_r268237180) [](ancestors = 268237180)

**Comment by jaredpar** [comment]
Done with review pass (iteration 5).

File: src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs (hunk 15 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs]: The idea was to make it resemble the property LocalAccessibility which indicates any accessibility mods directly on the accessor.

Comment (jaredpar) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs]: Okay. That makes sense. I'll take prior art over 8AM opinions.

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs]: I’ll add an “Is” though to better suit the bool return type.

File: src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs (hunk 11 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs]: Public API isn't exposed yet--we're also not positive about whether to expose `IsDeclaredReadOnly`, `IsEffectivelyReadOnly`, or both.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbol.cs (hunk 14 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbol.cs]: This implementation forbids making any auto-property or accessor `readonly`. Sounds like we don't have consensus on this design quite yet. Will revisit this once we do. @agocke

Comment (jaredpar) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbol.cs]: I feel like adding `readonly` to an auto-property is logically no different than adding `readonly` to any member of a `readonly struct`. Not sure how we could allow one without the other.

Additionally it's the act of making implicit behavior explicit. C# generally speaking allows this and in some cases, like `static class`, actually requires it.

File: src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs (hunk 13 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs]: What's currently done for accessibility modifiers is that instead of a code fix which "consolidates" the duplicate accessibility modifiers we just add yet another accessibility modifier. So, it seems like we could do better on both..

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 16]
```diff
@@ -5427,6 +5427,21 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
   <data name="WRN_ImplicitCopyInReadOnlyMember_Title" xml:space="preserve">
     <value>Call to non-readonly member from a 'readonly' member results in an implicit copy.</value>
   </data>
+  <data name="ERR_StaticMemberCantBeReadOnly" xml:space="preserve">
+    <value>Static member '{0}' cannot be 'readonly'.</value>
```
Comment (agocke) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: Let's give a little more info here, like: "readonly members cannot modify 'this' and static members do not have a 'this' reference"

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 17]
```diff
@@ -5427,6 +5427,21 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
...
+  </data>
+  <data name="ERR_AutoPropertyCantBeReadOnly" xml:space="preserve">
+    <value>Auto-implemented property or accessor '{0}' cannot be marked 'readonly'.</value>
+  </data>
+  <data name="ERR_InvalidPropertyReadOnlyMods" xml:space="preserve">
+    <value>Cannot specify 'readonly' modifiers on both property or indexer '{0}' and its accessors.</value>
```
Comment (agocke) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: This sentence doesn't parse right to me. I think it should say "Duplicate 'readonly' modifier on both the property and an accessor. Remove the modifier on the property or the modifier on the accessor."

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 18]
```diff
@@ -5427,6 +5427,21 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
...
+  </data>
+  <data name="ERR_DuplicatePropertyReadOnlyMods" xml:space="preserve">
+    <value>Cannot specify 'readonly' modifiers on both accessors of property or indexer '{0}'.</value>
```
Comment (agocke) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: I'd say how to fix this as well, namely put `readonly` on the whole property.

File: src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs (hunk 8 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs]: Done

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceEventSymbol.cs (hunk 9 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceEventSymbol.cs]: I contemplated this for a bit and for now I feel introducing nesting or locals in such a long if-else chain is going to be bad for clarity.

File: src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs (hunk 12 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs]: We're now going to allow `readonly` getter-only properties and `readonly` get accessors but there's no need to change this test back.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbol.cs (hunk 14 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbol.cs]: Done.

**Comment by RikkiGibson** [comment]
Pushed some improvements to error messages. Please take a look @agocke

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceEventSymbol.cs (hunk 9 above)
Comment (agocke) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceEventSymbol.cs]: I agree with @RikkiGibson
